source/fe/fe_values_views_internal.cc

```diff
@@ -0,0 +1,1167 @@
+                   ++q_point)
+                derivatives[q_point][comp] += value * (*shape_derivative_ptr++);
+            }
+          else
+            for (unsigned int d = 0; d < spacedim; ++d)
+              if (shape_function_data[shape_function]
+                    .is_nonzero_shape_function_component[d])
+                {
+                  const dealii::Tensor<order, spacedim> *shape_derivative_ptr =
+                    &shape_derivatives[shape_function_data[shape_function]
+                                         .row_index[d]][0];
+                  for (unsigned int q_point = 0; q_point < n_quadrature_points;
+                       ++q_point)
+                    derivatives[q_point][d] +=
+                      value * (*shape_derivative_ptr++);
+                }
+        }
+    }
+
+
+
+    template <int dim, int spacedim, typename Number>
+    void
+    do_function_symmetric_gradients(
+      const ArrayView<const Number> &              dof_values,
+      const Table<2, dealii::Tensor<1, spacedim>> &shape_gradients,
+      const std::vector<typename Vector<dim, spacedim>::ShapeFunctionData>
+        &shape_function_data,
+      std::vector<
+        typename ProductType<Number,
+                             dealii::SymmetricTensor<2, spacedim>>::type>
+        &symmetric_gradients)
+    {
+      const unsigned int dofs_per_cell       = dof_values.size();
+      const unsigned int n_quadrature_points = symmetric_gradients.size();
+
+      std::fill(
+        symmetric_gradients.begin(),
+        symmetric_gradients.end(),
+        typename ProductType<Number,
+                             dealii::SymmetricTensor<2, spacedim>>::type());
+
+      for (unsigned int shape_function = 0; shape_function < dofs_per_cell;
+           ++shape_function)
+        {
+          const int snc =
+            shape_function_data[shape_function].single_nonzero_component;
+
+          if (snc == -2)
+            // shape function is zero for the selected components
+            continue;
+
+          const Number &value = dof_values[shape_function];
+          // For auto-differentiable numbers, the fact that a DoF value is zero
+          // does not imply that its derivatives are zero as well. So we
+          // can't filter by value for these number types.
+          if (CheckForZero<Number>::value(value) == true)
+            continue;
+
+          if (snc != -1)
+            {
+              const unsigned int comp = shape_function_data[shape_function]
+                                          .single_nonzero_component_index;
+              const dealii::Tensor<1, spacedim> *shape_gradient_ptr =
+                &shape_gradients[snc][0];
+              for (unsigned int q_point = 0; q_point < n_quadrature_points;
+                   ++q_point)
+                symmetric_gradients[q_point] +=
+                  value * dealii::SymmetricTensor<2, spacedim>(
+                            symmetrize_single_row(comp, *shape_gradient_ptr++));
```

Comment:
I still think it's worth optimizing these functions because they are used a lot in our postprocessing / visualization infrastructure - that's something everyone uses.